CMakeLists.txt

```diff
@@ -25,50 +22,65 @@ set (SRCS
     "src/mixer_hw.c"
     "src/mixer_plugin.c")
 
-add_library("tinyalsa" ${HDRS} ${SRCS})
-target_compile_options("tinyalsa" PRIVATE -Wall -Wextra -Werror -Wfatal-errors ${plugin_opt})
-target_include_directories("tinyalsa" PRIVATE "include")
-
-macro(ADD_EXAMPLE EXAMPLE)
-    add_executable(${EXAMPLE} ${ARGN})
-    target_link_libraries(${EXAMPLE} "tinyalsa")
-    target_link_libraries(${EXAMPLE} "dl")
-    target_include_directories(${EXAMPLE} PRIVATE "include")
-endmacro(ADD_EXAMPLE EXAMPLE)
+set_property(TARGET "tinyalsa" PROPERTY PUBLIC_HEADER
+    "include/tinyalsa/attributes.h"
+    "include/tinyalsa/version.h"
+    "include/tinyalsa/asoundlib.h"
+    "include/tinyalsa/pcm.h"
+    "include/tinyalsa/plugin.h"
+    "include/tinyalsa/mixer.h")
 
-add_example("pcm-readi" "examples/pcm-readi.c")
-add_example("pcm-writei" "examples/pcm-writei.c")
+set_target_properties("tinyalsa" PROPERTIES
+    VERSION ${TinyALSA_VERSION}
+    SOVERSION ${TinyALSA_VERSION_MAJOR})
 
-macro(ADD_UTIL UTIL)
-    add_executable(${UTIL} ${ARGN})
-    target_link_libraries(${UTIL} "tinyalsa")
-    target_link_libraries(${UTIL} "dl")
-    target_compile_options(${UTIL} PRIVATE -Wall -Wextra -Werror -Wfatal-errors)
-    target_include_directories(${UTIL} PRIVATE "include")
-endmacro(ADD_UTIL UTIL)
+target_include_directories("tinyalsa" PUBLIC "include")
+target_compile_definitions("tinyalsa" PRIVATE
+    $<$<BOOL:${TINYALSA_USES_PLUGINS}>:TINYALSA_USES_PLUGINS>)
+target_link_libraries("tinyalsa" PUBLIC ${CMAKE_DL_LIBS})
 
-add_util("tinyplay" "utils/tinyplay.c")
-add_util("tinycap" "utils/tinycap.c")
-add_util("tinypcminfo" "utils/tinypcminfo.c")
-add_util("tinymix" "utils/tinymix.c")
+# Examples
+if(TINYALSA_BUILD_EXAMPLES)
+    set(TINYALSA_EXAMPLES pcm-readi pcm-writei)
+else()
+    set(TINYALSA_EXAMPLES)
+endif()
 
-install(FILES ${HDRS}
-    DESTINATION "include/tinyalsa")
+foreach(EXAMPLE IN LISTS TINYALSA_EXAMPLES)
+    add_executable("${EXAMPLE}" "examples/${EXAMPLE}.c")
+    target_link_libraries("${EXAMPLE}" PRIVATE "tinyalsa")
+endforeach()
 
-if(NOT DEFINED CMAKE_INSTALL_LIBDIR)
-    set(CMAKE_INSTALL_LIBDIR lib)
+# Utilities
+if(TINYALSA_BUILD_UTILS)
+    set(TINYALSA_UTILS tinyplay tinycap tinypcminfo tinymix)
+else()
+    set(TINYALSA_UTILS)
 endif()
 
-if(NOT DEFINED CMAKE_INSTALL_BINDIR)
-	set(CMAKE_INSTALL_BINDIR bin)
-endif()
+foreach(UTIL IN LISTS TINYALSA_UTILS)
+    add_executable("${UTIL}" "utils/${UTIL}.c")
+    target_link_libraries("${UTIL}" PRIVATE "tinyalsa")
+endforeach()
 
-install(TARGETS "tinyalsa"
-                "tinyplay"
-                "tinycap"
-                "tinymix"
-                "tinypcminfo"
+# Add C warning flags
+include(CheckCCompilerFlag)
+foreach(FLAG IN ITEMS -Wall -Wextra -Werror -Wfatal-errors)
```

Comment:
I did it just because I view it as good practice to check for support for flags before adding them, especially when they're stuff like warning flags that don't add crucial functionality, to avoid any cases potentially coming up where a compiler doesn't support them. In my opinion that's better than special-casing certain compilers and proactively avoids any issues that could arise here.